### ISSUE TITLE
Slow module initialization time on low-end devices

### ISSUE THREAD

**Issue by gaearon** [formatjs/formatjs#4463]
## Which package?

`@formatjs/intl-pluralrules/polyfill`

## Describe the bug

While not a bug per se, merely importing `@formatjs/intl-pluralrules/polyfill` causes a *one second stall* on a low-end Android device in a React Native project. For context, this is ~1s out of ~3.5s of our startup time, i.e. this polyfill is almost a third of the entire app's initialization time (which uses many other libraries including RN core itself).

I'm not sure what's special about this polyfill in particular. Maybe it's hitting some bad case. From inspecting the code, one thing that jumps out to me is that it pulls in the entirety of `@formatjs/ecma402-abstract` because it targets CJS, so a bunch of code gets initialized just to get `invariant`, `FormatNumericToString`, and `ToNumber`. Maybe publishing an ESM version would avoid that. 

## To Reproduce

This is reproducible with the [Bluesky app](https://github.com/bluesky-social/social-app/) (fully open source) on Samsung Galaxy A12. If there is interest, I can give more detailed instructions, but it's essentially about installing it, running `yarn`, then `yarn android` to run it on an Android simulator (assuming you have one set up).

**Comment by gaearon** [comment]
It seems like the issue is that the best fit matcher is very inefficient. You can see this by running the polyfill code unconditionally in Chrome with 6x CPU throttling.

![Screenshot 2024-06-18 at 15 59 40](https://github.com/formatjs/formatjs/assets/810438/5afdc977-8995-46e3-a6bf-35f677483099)

**Comment by longlho** [comment]
yeah best fit matching is non-trivial. The algorithm is in LDML [here](https://unicode.org/reports/tr35/#language-matching). At a high level it has to calculate the distance between the requested locale and the dataset, taken into account some special regional grouping & preference like Latin America, then take the shortest one. We've already put in some shortcuts wrt local distance calculation, but overall it's quite heavy. Getting this to match icu4j is pretty tricky. I can take another look, but no promises.

**Comment by gaearon** [comment]
Cool, thanks. Would it be possible to offer a version of the polyfill that does `'lookup'` instead?

Also, in general, something feels pathological about, say, passing `en` and then spending so much initialization time doing something. Zooming out of the specific CPU work being done here — is there any way something could be precomputed here? How do I explain it to myself why this CPU work needs to happen during initialization? Or maybe, could any of this work be shifted towards when it's absolutely necessary?

**Comment by longlho** [comment]
yeah because the 1st thing it tries to do is detect whether this platform needs polyfilling, which checks for a specific Chrome bug. If you're running on a platform that you know definitely needs polyfilling, you can import `/polyfill-force` instead.

The `lookup` locale algorithm is pretty bad, as it basically just removes 1 subtag at a time when trying to match, which would  yield false negative for things like traditional Chinese.

**Comment by gaearon** [comment]
>If you're running on a platform that you know definitely needs polyfilling, you can import /polyfill-force instead.

Fantastic, this hasn't even occurred to me but it's exactly what I needed.

**Comment by gaearon** [comment]
It would be still be great to _somehow_ get this fixed upstream because I bet it's affecting a lot of RN apps that use `@formatjs` transitively. Most developers aren't aware how to debug init time issues so it probably translates to second stalls across a lot of apps.

**Comment by gaearon** [comment]
Actually nvm, I guess developers *do* include the polyfill itself manually so this was probably our mistake. Would be good to call it out in docs maybe.

**Comment by gaearon** [comment]
Thanks a lot for holding my hand through this, the last insight let me cut off a few hundred more ms.

**Comment by sregg** [comment]
We improved our app start time by 3x by just replacing all polyfill imports with their `-force` versions.
That's 2s instead of 7s on my Pixel 6 Pro.
Thanks @gaearon for the hint 🙏

**Comment by gaearon** [comment]
Might be dicey but seems like a good moment to kick off a discussion on how to avoid this slowing down more apps. It sucks that many apps will add this polyfill without testing perf and the users will be stuck with bad startup perf as a result.

https://github.com/formatjs/formatjs/pull/4485

**Comment by gzurbach** [comment]
Thank you for this incredible helpful advice. Switching all polyfill imports to their `-force` versions was the single most impactful change for Android cold start performance:

<img width="782" height="373" alt="Image" src="https://github.com/user-attachments/assets/d6eceb87-c40c-41c7-97d3-d3068d9753a1" />

**Comment by longlho** [comment]
we landed some major perf gain here so closing this properly

**Comment by rodperottoni** [comment]
@longlho can you reference a PR / hash of said major perf gain? Thanks!

**Comment by longlho** [comment]
yup https://github.com/formatjs/formatjs/pull/5674

The README has more details as well https://github.com/formatjs/formatjs/blob/main/packages/intl-localematcher/README.md#three-tier-optimization
